src/undo-manager.js

```diff
@@ -18,70 +20,80 @@ define(function () {
 
     transaction.execute();
 
-    this._stack.splice(0, this.position);
-    if (merge && this.length) {
-      this._stack[0].push(transaction);
-    }
-    else {
-      this._stack.unshift([transaction]);
+    if (this.position > 0) {
+      this.clearRedo();
     }
-    this.position = 0;
 
-    if (this._limit && this._stack.length > this._limit) {
-      this.length = this._stack.length = this._limit;
+    var transactions;
+    if (merge && this.length) {
+      transactions = this._stack.first().push(transaction);
+      this._stack = this._stack.shift().unshift(transactions);
     }
     else {
-      this.length = this._stack.length;
-    }
+      transactions = Immutable.List.of(transaction);
+      this._stack = this._stack.unshift(transactions);
+      this.length++;
 
-    if (this._fireEvent) {
-      this._ush.dispatchEvent(new CustomEvent('DOMTransaction', {detail: {transactions: this._stack[0].slice()}, bubbles: true, cancelable: false}));
+      if (this._limit && this.length > this._limit) {
+        this.clearUndo(this._limit);
+      }
     }
+
+    this._dispatch('DOMTransaction', transactions);
   };
 
   UndoManager.prototype.undo = function () {
-    if (this.position < this.length) {
-      for (var i = this._stack[this.position].length - 1; i >= 0; i--) {
-        this._stack[this.position][i].undo();
-      }
-      this.position++;
+    if (this.position >= this.length) { return; }
 
-      if (this._fireEvent) {
-        this._ush.dispatchEvent(new CustomEvent('undo', {detail: {transactions: this._stack[this.position - 1].slice()}, bubbles: true, cancelable: false}));
-      }
+    var transactions = this._stack.get(this.position);
+    var i = transactions.size;
+    while (i--) {
+      transactions.get(i).undo();
     }
+    this.position++;
+
+    this._dispatch('undo', transactions);
   };
 
   UndoManager.prototype.redo = function () {
-    if (this.position > 0) {
-      for (var i = 0, n = this._stack[this.position - 1].length; i < n; i++) {
-        this._stack[this.position - 1][i].redo();
-      }
-      this.position--;
+    if (this.position === 0) { return; }
 
-      if (this._fireEvent) {
-        this._ush.dispatchEvent(new CustomEvent('redo', {detail: {transactions: this._stack[this.position].slice()}, bubbles: true, cancelable: false}));
-      }
+    this.position--;
+    var transactions = this._stack.get(this.position);
+    var i = -1;
+    while (i++ < transactions.size - 1) {
+      transactions.get(i).redo();
     }
+
+    this._dispatch('redo', transactions);
   };
 
   UndoManager.prototype.item = function (index) {
-    if (index >= 0 && index < this.length) {
-      return this._stack[index].slice();
-    }
-    return null;
+    return index >= 0 && index < this.length ?
+      this._stack.get(index).toArray() :
+      null;
```

Comment:
I know this is what the code used to do but would it be better if it was `undefined` that was returned at this point?